File: problem2.java
package Stack-2;

public class problem2 {
    class Solution {
        //TC:- O(n)
        //SC:- O(n)
        public boolean isValid(String s) {
            Stack<Character>st = new Stack<>();
            
            for(int i=0; i<s.length(); i++){
                char c = s.charAt(i);
                
                if(c=='}'){
                    if(!st.isEmpty()&& st.peek()=='{'){
                        st.pop();
                    }else{
                        st.add('}');
                    }
                } else if(c==']'){
                    if(!st.isEmpty() && st.peek()=='['){
                        st.pop();
                    }else{
                        st.add(']');
                    }
                }else if(c==')'){
                    if(!st.isEmpty() && st.peek()=='('){
                        st.pop();
                    }else{
                        st.add(')');
                    }
                }else{
                st.add(c);
                }
            }
            
            if(st.isEmpty()){return true;}
            return false;
        }
    }
}

File: Problem_2.js
// ## Problem2 Valid Parentheses (https://leetcode.com/problems/valid-parentheses/)

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
    if (s === null || s.length === 0)
        return true;
    let stack = [];
    for (let i = 0; i < s.length; i++) {
        let ch = s[i];
        // If found an opening brace, push its equivalent closing brace to stack 
        if (ch === "{") {
            stack.push("}");
        } else if (ch === "(") {
            stack.push(")");
        } else if (ch === "[") {
            stack.push("]");
        } else {
            // Check if the closing brace is same as the last opening brace
            if (stack.length === 0 || stack.pop() !== ch) {
                return false;
            }
        }
    }
    // If there are still braces which have not been closed, return false
    return stack.length === 0;
};
